ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala

```diff
@@ -38,19 +45,35 @@ class ApiPackageManagementService(
               pkgId.toString,
               details.size,
               Some(Timestamp(details.knownSince.getEpochSecond, details.knownSince.getNano)),
-              details.sourceDescription)
+              details.sourceDescription.getOrElse(""))
         })
       }(DE)
   }
 
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] = {
-    FutureConverters
-      .toScala(packagesWrite.uploadDar("", request.darFile.toByteArray))
-      .flatMap {
-        case UploadDarResult.Ok => Future.successful(UploadDarFileResponse())
-        case UploadDarResult.InvalidPackage(err) =>
-          Future.failed(ErrorFactories.invalidArgument(s"Invalid package: $err"))
-      }(DE)
+    val resultT = for {
+      file <- Try(File.createTempFile("uploadDarFile", ".dar"))
+      _ <- bracket(Try(new FileOutputStream(file)))(fos => Try(fos.close())).flatMap { fos =>
+        Try(fos.write(request.darFile.toByteArray))
+      }
+      dar <- DarReader { case (_, x) => Try(Archive.parseFrom(x)) }.readArchive(new ZipFile(file))
```

Comment:
I think it would be nice to have `DarReader` support `ZipInputStream` as well if possible. This way we don't have to create a temporary zipfile, but can feed the `ZipInputStream` directly from the byte array.

Reply:
I agree this would be better. The `DarReader` does random-access reading of zip file entries, which does not work with `ZipInputStream`, so this would require some refactoring of the reader.

Reply:
Right. I didn't mean for this to be changed in this PR. Thanks for checking.